### PR TITLE
198 as a participant i want to be provided gps navigation to the event

### PR DESCRIPTION
**Pull request by Mathurinnn**
the map has never been better; now features a dynamic fetching of events based on the position of the camera on a 1km x 1km square, and removes the displaying and fetching of events if the zoom is too low.
Clicking on an event displays the title of the event and a minimal description of the event. Clicking again redirects to the event's own screen where we can register and see all infos. As it is a GoogleMap composable, clicking on an event displays an icon to go the event via the Maps app.
The user's position, events the user is registered to and other events are also displayed using different pin icons.
The map is usable

### CODE REVIEW

**Comment by Mathurinnn** [comment]
Ok so this lacks some tests, I propose I fix coverage in next PR by tomorrow, and we merge this now

File: app/src/main/java/com/github/se/gatherspot/model/MapViewModel.kt [hunk 1]
```diff
@@ -18,25 +19,81 @@ import com.github.se.gatherspot.model.event.Event
...
+import kotlin.math.PI
+import kotlin.math.cos
 
 @RequiresApi(Build.VERSION_CODES.S)
 class MapViewModel(application: Application) : AndroidViewModel(application) {
 
+  companion object {
+    fun degreeToMeters(latitudeDegrees: Double): Double {
+      // Radius of the Earth at the equator in meters
+      val earthRadiusAtEquator = 6378137.0 // in meters
```
Comment (markh0rr) [inline app/src/main/java/com/github/se/gatherspot/model/MapViewModel.kt:32]: you can write this as EARTH_RADIUS_AT_EQUATOR to follow the constant naming norm

File: app/src/main/java/com/github/se/gatherspot/model/MapViewModel.kt [hunk 2]
```diff
@@ -18,25 +19,81 @@ import com.github.se.gatherspot.model.event.Event
...
+      // Where `latitude` is in radians and `radius` is the radius at that latitude
+      val radiusAtLatitude = earthRadiusAtEquator * cos(latitudeRadians)
+      val lengthOfDegree = (2 * PI * radiusAtLatitude) / 360.0
+
+      return lengthOfDegree
+    }
+
+    fun metersToDegree(latitudeDegrees: Double, meters: Double): Double {
+      // Radius of the Earth at the equator in meters
+      val earthRadiusAtEquator = 6378137.0 // in meters
```
Comment (markh0rr) [inline app/src/main/java/com/github/se/gatherspot/model/MapViewModel.kt:49]: as earthRadiusAtEquator is used in 2 methods you may put it as a constant of the companion object

File: app/src/main/java/com/github/se/gatherspot/model/MapViewModel.kt [hunk 3]
```diff
@@ -18,25 +19,81 @@ import com.github.se.gatherspot.model.event.Event
...
+
+      // Convert latitude from degrees to radians
+      val latitudeRadians = Math.toRadians(latitudeDegrees)
+
+      // Calculate the length of a degree of latitude in meters
+      // This accounts for the Earth's curvature using the WGS-84 ellipsoid
+      // Formula: length_of_degree = (2 * PI * radius) * (cos(latitude))
+      // Where `latitude` is in radians and `radius` is the radius at that latitude
+      val radiusAtLatitude = earthRadiusAtEquator * cos(latitudeRadians)
```
Comment (markh0rr) [inline app/src/main/java/com/github/se/gatherspot/model/MapViewModel.kt:58]: same for radiusAtLatitude and lengthOfDegree

File: app/src/main/java/com/github/se/gatherspot/firebase/EventFirebaseConnection.kt [hunk 4]
```diff
@@ -280,6 +280,27 @@ class EventFirebaseConnection : FirebaseConnectionInterface<Event> {
     return eventsFromQuerySnapshot(querySnapshot)
   }
 
+  suspend fun fetchAllInPerimeter(
+      latitude: Double,
+      longitude: Double,
+      degrees: Double
+  ): MutableList<Event> {
+
+    val querySnapshot: QuerySnapshot =
```
Comment (markh0rr) [inline app/src/main/java/com/github/se/gatherspot/firebase/EventFirebaseConnection.kt:289]: this is a great implementation to not fetch all events but only the events nearby

**Review by markh0rr — CHANGES_REQUESTED** [review]
a good start for the map and display  of nearby events

**Comment by Mathurinnn** [comment]
thanks for the review, these problems will be solved in pr#219

**Pull request merged**